inference-engine/src/inference_engine/include/openvino/runtime/core.hpp

```diff
@@ -0,0 +1,308 @@
+     *
+     * @return A vector of devices. The devices are returned as { CPU, FPGA.0, FPGA.1, MYRIAD }
+     * If there more than one device of specific type, they are enumerated with .# suffix.
+     */
+    std::vector<std::string> get_available_devices() const;
+
+    /**
+     * @brief Register new device and plugin which implement this device inside Inference Engine.
+     *
+     * @param pluginName A name of plugin. Depending on platform pluginName is wrapped with shared library suffix and
+     * prefix to identify library full name
+     *
+     * @param deviceName A device name to register plugin for. If device name is not specified, then it's taken from
+     * plugin itself.
+     */
+    void register_plugin(const std::string& pluginName, const std::string& deviceName);
+
+    /**
+     * @brief Unloads previously loaded plugin with a specified name from Inference Engine
+     * The method is needed to remove plugin instance and free its resources. If plugin for a
+     * specified device has not been created before, the method throws an exception.
+     *
+     * @param deviceName Device name identifying plugin to remove from Inference Engine
+     */
+    void unload_plugin(const std::string& deviceName);
+
+    /** @brief Registers plugin to Inference Engine Core instance using XML configuration file with
+     * plugins description.
+     *
+     *  XML file has the following structure:
+     *
+     * ```xml
+     * <ie>
+     *     <plugins>
+     *         <plugin name="" location="">
+     *             <extensions>
+     *                 <extension location=""/>
+     *             </extensions>
+     *             <properties>
+     *                 <property key="" value=""/>
+     *             </properties>
+     *         </plugin>
+     *     </plugins>
+     * </ie>
+     * ```
+     *
+     * - `name` identifies name of device enabled by plugin
+     * - `location` specifies absolute path to dynamic library with plugin. A path can also be relative to inference
+     * engine shared library. It allows to have common config for different systems with different configurations.
+     * - Properties are set to plugin via the `SetConfig` method.
+     * - Extensions are set to plugin via the `AddExtension` method.
+     *
+     * @param xmlConfigFile A path to .xml file with plugins to register.
+     */
+    void register_plugins(const std::string& xmlConfigFile);
+
+    /**
+     * @brief Create a new shared context object on specified accelerator device
+     * using specified plugin-specific low level device API parameters (device handle, pointer, etc.)
+     * @param deviceName Name of a device to create new shared context on.
+     * @param params Map of device-specific shared context parameters.
+     * @return A shared pointer to a created remote context.
+     */
+    InferenceEngine::RemoteContext::Ptr create_context(const std::string& deviceName, const InferenceEngine::ParamMap& params);
```

Comment:
If we pass something using `ptr*` in the header let's use the forward declaration as match as possible.  Let's use forward declared `InferenceEngine::RemoteContext` and `ngraph::Function`

Reply:
same for `IExtensionPtr`

Reply:
And if it is possible for `Blob::Ptr`

Reply:
Certainly, we should remove headers `#includes` with this declaration.

Reply:
Fixed